Private Event Project

haraphat01 opened the pull request:
This is my final project for private events, on the rails curriculum. This project tests my understanding of the usage of Active Querries.

To project is just a simple event page, where a user can create as many events as possible, the events created by any user can be seen at the bottom of the page, showing the location and the date of the event.

Kindly check for your kind approval.

haraphat01 committed:
Added the registration part

haraphat01 committed:
Created the model for the app

haraphat01 committed:
Created the controller for the app

haraphat01 committed:
Working on the functionality of the page

haraphat01 committed:
Fixed the linter errors

haraphat01 committed:
Added rspec

haraphat01 committed:
Fixed lint errors

haraphat01 committed:
Fixed linter errors

haraphat01 committed:
fixed some indentations on the  hmtl

mcrd25 commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -0,0 +1,30 @@
+
+<%= form_for(resource, as: resource_name, url: registration_path(resource_name)) do |f| %>
+  <%= render "devise/shared/error_messages", resource: resource %>
+
+  <div class="field">
+    <%= f.label :email %><br />
+    <%= f.email_field :email, autofocus: true, autocomplete: "email" %>
+  </div>
+  
+
+  <div class="field">
+    <%= f.label :password %>
+    <% if @minimum_password_length %>
+    <em>(<%= @minimum_password_length %> characters minimum)</em>
+    <% end %><br />
+    <%= f.password_field :password, autocomplete: "new-password" %>
+  </div>
+
+  <div class="field">
+    <%= f.label :password_confirmation %><br />
+    <%= f.password_field :password_confirmation, autocomplete: "new-password" %>
+  </div>
+
+  <div class="actions">
+    <%= f.submit "Sign up" %>
+  </div>
+<% end %>
```
I am told that I'm missing username field but I don't see one here.
![image](https://user-images.githubusercontent.com/31193823/90804766-a8c51600-e2e8-11ea-86c4-d731a78dba66.png)

I can't really test the application otherwise

mcrd25 commented on app/models/user.rb:
```diff
@@ -0,0 +1,13 @@
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
+
+  has_many :events, foreign_key: 'creator_id', class_name: 'Event'
```
`class_name` is not needed here. class should be configured on its own given the name used is `:events` had it been different, then yes

mosesogwo replied:
See the previous reviewer's comment on this. class_name is not needed here.

mcrd25 commented on app/controllers/events_controller.rb:
```diff
@@ -0,0 +1,40 @@
+    @user = User.find(@event.creator_id)
+  end
+
+  def past
+    Event.past
+  end
+
+  def upcoming
+    Event.upcoming
+  end
```
are these necessary? It'd be better to have `@upcoming = Event.upcoming` etc. in `index`

mcrd25 reviewed (changes requested):
# changes required :mag: 

Hi @haraphat01 :wave: 

Good attempt. :slightly_smiling_face: 

However, I cannot test the application itself due to the inability to register successfully. (See my comment)

Other than that issue here are things you can work on before submitting re-review:
#### required changes
- You need more tests. don't forget you ought to have unit and integration tests. You currently only have 3 tests and they all seem to fail on my machine.
  - unit tests for validations and associations [see screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/unit_tests_requirement.png) pathwright
  - authentication have has integration tests [see screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png)
  - event management features have an integration test. [see screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png)
- Address any review comments that are not marked as optional


Hope my review was helpful!

***Please makes changes and submit another review request.
If you have any questions, let me know here or on slack and I will be glad to clarify with you. You can also reach me and connect with me on my social media or email.***

*Happy coding 😄,*
**[--Maya Douglas](https://github.com/mcrd25)**
[GitHub](https://github.com/mcrd25) [Twitter](https://twitter.com/mcrd25) [linkedin](https://www.linkedin.com/in/mayadouglas/) [GitLab](https://gitlab.com/mcrd25)

---
*P.S. all requested changes not marked by [OPTIONAL] or are not positive comments are **REQUIRED***

haraphat01 committed:
Adding the test files

haraphat01 committed:
Worked on the TSE review

haraphat01 committed:
Worked on the TSE review

haraphat01 committed:
Added @ to the event controller

haraphat01 commented:
Thank you for the review, I have visited the pages that need changes and amended it.  

- You would now be able to sign up.

- I added the username row was missing initially

- I reworked on the test files

Kindly check for your kind approval.

mosesogwo commented on app/views/users/show.html.erb:
```diff
@@ -0,0 +1,30 @@
+<h2>Events you are attending:</h2>
+<h3>Upcoming events:</h3>
+<table>
+	<%= render "events/event", obj: @user.attended_events.upcoming %>
```
You could retrieve this collection _"@user.attended_events.upcoming"_ into a variable from the controller just like you have done for created events above.

mosesogwo commented on app/views/users/show.html.erb:
```diff
@@ -0,0 +1,30 @@
+</table>
+<h3>Past events:</h3>
+<table>
+	<%= render "events/event", obj: @user.attended_events.past %>
```
You could retrieve this collection _"@user.attended_events.past"_ into a variable from the controller just like you have done for created events above.

mosesogwo reviewed (changes requested):
**Hello!:wave: Nice improvements :+1:  Just a few more changes and you'll be good to :rocket:**

:point_right: Add indices to foreign keys on Events and Invites table.
You can read about why this is important here:
https://semaphoreci.com/blog/2017/05/09/faster-rails-is-your-database-properly-indexed.html
https://sqlperformance.com/2012/11/t-sql-queries/benefits-indexing-foreign-keys

:point_right: Test negative scenario for your signup feature.

:point_right: Extract all view logic into helper methods. Check your application layout, event index and partial for this.

_**Kindly make these changes and submit another code review request.**_

**Any questions? You can reach me on Slack (Moses Ogwo)**

**Cheers:stars:**

### Status: Changes Requested:exclamation:

haraphat01 committed:
Index added to both event and invites

haraphat01 committed:
Fixed the failed test

haraphat01 committed:
All my logic arranged

haraphat01 committed:
Fixed lint errors

haraphat01 commented:
Thanks for the review, here are the changes made so far.

- Indices added to both Event and Invite model

-  Failed test corrected

- My logic has been reorganized.

wathigo reviewed (changes requested):
### Changes Required.
Great work refactoring the code.

**REQUIRED**
* We have one failing test. See :point_down: 
![Screenshot from 2020-08-22 11-20-21](https://user-images.githubusercontent.com/44072711/90952111-99b9a180-e469-11ea-9c08-fc2b5216fde4.png)

**After working on the changes, submit another request for review.**

**Happy Coding :computer: 
Simon Wathigo.**

haraphat01 committed:
Fixed the failed test

haraphat01 commented:
Thank you for your review.

The failed test has been fixed, this came up after refactoring the codes, but it's working now.

maelfosso reviewed (approved):
# Congrtulations :clap: 

THank you for applying the suggested changes.
You can merge it now :+1: 

Good continuation :handshake: 

Mael FOSSO
